### PR TITLE
Update to Saleor 3 and shipping

### PR DESCRIPTION
**Pull request by bufke**
# Changes

- Port to Saleor 3.0 (again)
- Update test fixtures
- Add "Shipping Product Code" as configurable value
- Add support for optional shipping tax calculation
- Better log message for plugin authentication error

# Working on

- Trying to get demo account from Avalara so that tests are more reproducible
- Unit tests need improvement

# Questions

- How do I run tests on this? Currently I run tests by mounting my plugin volume into a saleor 3 repo. There must be a better way. Looks like CI is failing too. That said I'm able to run tests locally just fine so it's only a problem when committing here.

### CODE REVIEW

**Comment by korycins** [comment]
`How do I run tests on this? Currently I run tests by mounting my plugin volume into a saleor 3 repo. There must be a better way. Looks like CI is failing too. That said I'm able to run tests locally just fine so it's only a problem when committing here.` Looks like the CI is passing in your PR. It just simple download the current master of Saleor and run tests by using its interface.

**Review by korycins — APPROVED** [review]
Looks good, but could you run black and isort on the code also? (not sure if you already did that). You can use Saleor's configuration for that. 
It would be easier for me to include black & isort in the future. 
Also I see a lot of unneeded imports, can we remove them? (plugin.py, utils.py, test file)

File: excise/plugin.py [hunk 1]
```diff
@@ -133,33 +161,26 @@ def calculate_checkout_total(
             return previous_value
         checkout_total = previous_value
 
-        if not _validate_checkout(checkout, [line_info.line for line_info in lines]):
+        if not _validate_checkout(checkout_info, lines):
             logger.debug("Checkout Invalid in Calculate Checkout Total")
             return checkout_total
 
-        response = get_checkout_tax_data(checkout, discounts, self.config)
+        checkout = checkout_info.checkout  ## change this
```
Comment (fowczarek) [inline excise/plugin.py]: I don't understand this comment. Can you add more info if this comment is necessary?

**Comment by korycins** [comment]
@bufke  Would be also great if you could run flake8 on the code. 
Here is the report:
```
./excise/tasks.py:17:80: E501 line too long (81 > 79 characters)
./excise/tasks.py:24:80: E501 line too long (81 > 79 characters)
./excise/tasks.py:29:80: E501 line too long (86 > 79 characters)
./excise/tests/test_avatax_excise.py:4:1: F401 'urllib.parse.urljoin' imported but unused
./excise/tests/test_avatax_excise.py:26:1: F811 redefinition of unused 'AvataxConfiguration' from line 25
./excise/tests/test_avatax_excise.py:26:1: F401 '..utils.get_api_url' imported but unused
./excise/tests/test_avatax_excise.py:34:1: F811 redefinition of unused 'fetch_checkout_lines' from line 18
./excise/tests/test_avatax_excise.py:108:80: E501 line too long (80 > 79 characters)
./excise/tests/test_avatax_excise.py:115:80: E501 line too long (80 > 79 characters)
./excise/tests/test_avatax_excise.py:116:80: E501 line too long (81 > 79 characters)
./excise/tests/test_avatax_excise.py:140:80: E501 line too long (81 > 79 characters)
./excise/tests/test_avatax_excise.py:164:80: E501 line too long (81 > 79 characters)
./excise/tests/test_avatax_excise.py:180:80: E501 line too long (88 > 79 characters)
./excise/tests/test_avatax_excise.py:194:80: E501 line too long (86 > 79 characters)
./excise/tests/test_avatax_excise.py:227:80: E501 line too long (86 > 79 characters)
./excise/tests/test_avatax_excise.py:250:80: E501 line too long (86 > 79 characters)
./excise/tests/test_avatax_excise.py:282:80: E501 line too long (86 > 79 characters)
./excise/tests/test_avatax_excise.py:296:80: E501 line too long (1670 > 79 characters)
./excise/tests/test_avatax_excise.py:302:80: E501 line too long (83 > 79 characters)
./excise/tests/test_avatax_excise.py:307:80: E501 line too long (86 > 79 characters)
./excise/tests/test_avatax_excise.py:354:80: E501 line too long (83 > 79 characters)
./excise/tests/test_avatax_excise.py:356:80: E501 line too long (86 > 79 characters)
./excise/tests/test_avatax_excise.py:369:80: E501 line too long (86 > 79 characters)
./excise/tests/test_avatax_excise.py:402:80: E501 line too long (86 > 79 characters)
./excise/tests/test_avatax_excise.py:414:80: E501 line too long (86 > 79 characters)
./excise/tests/test_avatax_excise.py:422:80: E501 line too long (84 > 79 characters)
./excise/tests/test_avatax_excise.py:427:80: E501 line too long (86 > 79 characters)
./excise/tests/test_avatax_excise.py:460:80: E501 line too long (86 > 79 characters)
./excise/tests/test_avatax_excise.py:522:80: E501 line too long (83 > 79 characters)
./excise/tests/test_avatax_excise.py:530:80: E501 line too long (86 > 79 characters)
./excise/tests/test_avatax_excise.py:550:80: E501 line too long (86 > 79 characters)
./excise/tests/test_avatax_excise.py:555:80: E501 line too long (86 > 79 characters)
./excise/tests/test_avatax_excise.py:570:80: E501 line too long (83 > 79 characters)
./excise/tests/test_avatax_excise.py:605:80: E501 line too long (81 > 79 characters)
./excise/tests/test_avatax_excise.py:616:80: E501 line too long (86 > 79 characters)
./excise/tests/test_avatax_excise.py:659:80: E501 line too long (86 > 79 characters)
```

File: excise/tests/cassettes/test_order_created_wrong_data.yaml [hunk 2]
```diff
@@ -1,22 +1,465 @@
 interactions:
 - request:
-    body: '{"EffectiveDate": "02/24/21", "InvoiceDate": "02/24/21", "TitleTransferCode":
-      "DEST", "TransactionType": "RETAIL", "TransactionLines": [{"InvoiceLine": 70,
-      "ProductCode": "SKU_AA", "UnitPrice": "10.000", "UnitOfMeasure": null, "BilledUnits":
-      "3", "LineAmount": "12.300", "AlternateUnitPrice": "1.000", "TaxIncluded": true,
-      "UnitQuantity": null, "UnitQuantityUnitOfMeasure": null, "DestinationCountryCode":
-      "USA", "DestinationJurisdiction": "TX", "DestinationAddress1": "1100 Congress
-      Ave", "DestinationAddress2": "", "DestinationCounty": "", "DestinationCity":
+    body: '{"EffectiveDate": "2021-03-12", "InvoiceDate": "2021-03-12", "TitleTransferCode":
```
Comment (fowczarek) [inline excise/tests/cassettes/test_order_created_wrong_data.yaml:3]: This cassette has 6 requests. It's potentially a bug.

File: excise/tests/cassettes/test_preprocess_order_creation.yaml [hunk 3]
```diff
@@ -1,10 +1,10 @@
 interactions:
 - request:
-    body: '{"EffectiveDate": "02/18/21", "InvoiceDate": "02/18/21", "TitleTransferCode":
-      "DEST", "TransactionType": "RETAIL", "TransactionLines": [{"InvoiceLine": 9,
-      "ProductCode": "123", "UnitPrice": "10.000", "UnitOfMeasure": "PAC", "BilledUnits":
+    body: '{"EffectiveDate": "2021-05-10", "InvoiceDate": "2021-05-10", "TitleTransferCode":
```
Comment (fowczarek) [inline excise/tests/cassettes/test_preprocess_order_creation.yaml:3]: This cassette has 2 requests. Are we sure that we should send 2 requests to ATE in this function?

File: excise/tests/test_avatax_excise.py [hunk 4]
```diff
@@ -230,6 +240,63 @@ def test_calculate_checkout_line_total(
...
+
+    manager.calculate_checkout_line_total(
+        checkout_info,
+        lines,
+        checkout_line_info,
+        checkout_with_item.shipping_address,
+        discounts,
+    )
+
+    checkout = Checkout.objects.filter(token=line.checkout.token).first()
```
Comment (fowczarek) [inline excise/tests/test_avatax_excise.py:296]:
```suggestion
    checkout_with_item.refresh_from_db()
```

File: excise/tests/test_avatax_excise.py [hunk 5]
```diff
@@ -230,6 +240,63 @@ def test_calculate_checkout_line_total(
...
...AS", "Currency": "USD", "SubtotalInd": "C", "StatusCode": "ACTIVE", "QuantityInd": "B"}, {"TransactionTaxAmounts": [], "SequenceId": 2, "TransactionLine": 1, "InvoiceLine": 1, "CountryCode": "USA", "Jurisdiction": "TX", "LocalJurisdiction": "05000", "ProductCategory": 0.0, "TaxingLevel": "CIT", "TaxType": "S", "RateType": "G", "RateSubtype": "NONE", "CalculationTypeInd": "P", "TaxRate": 0.01, "TaxQuantity": 0.0, "TaxAmount": 0.28, "TaxExemptionInd": "N", "SalesTaxBaseAmount": 27.71, "LicenseNumber": "", "RateDescription": "TX CITY TAX - AUSTIN", "Currency": "USD", "SubtotalInd": "C", "StatusCode": "ACTIVE", "QuantityInd": "B"}, {"TransactionTaxAmounts": [], "SequenceId": 3, "TransactionLine": 1, "InvoiceLine": 1, "CountryCode": "USA", "Jurisdiction": "TX", "LocalJurisdiction": "6000814", "ProductCategory": 0.0, "TaxingLevel": "STJ", "TaxType": "S", "RateType": "G", "RateSubtype": "NONE", "CalculationTypeInd": "P", "TaxRate": 0.01, "TaxQuantity": 0.0, "TaxAmount": 0.28, "TaxExemptionInd": "N", "SalesTaxBaseAmount": 27.71, "LicenseNumber": "", "RateDescription": "TX SPECIAL TAX - AUSTIN MTA TRANSIT", "Currency": "USD", "SubtotalInd": "C", "StatusCode": "ACTIVE", "QuantityInd": "B"}]'
```
Comment (fowczarek) [inline excise/tests/test_avatax_excise.py:299]: Are you sure that we need to store all this data?

File: excise/utils.py [hunk 6]
```diff
@@ -0,0 +1,554 @@
...
+    variant: "ProductVariant",
+    shipping_address: "Address",
+    variant_channel_listing: "ProductVariantChannelListing",
+):
+    """Abstract line data regardless of Checkout or Order."""
+    stock = variant.stocks.for_country_and_channel(
+        shipping_address.country, variant_channel_listing.channel.slug
+    ).first()
+    if stock:
+        warehouse_address = stock.warehouse.address
```
Comment (fowczarek) [inline excise/utils.py:234]: I'm not sure that we should fetch the address for each line separately? It cost us 1 DB request per line.

File: excise/utils.py [hunk 7]
```diff
@@ -0,0 +1,554 @@
...
+    shipping_address = order.shipping_address
+    if shipping_address is None:
+        raise TaxError("Shipping address required for ATE tax calculation")
+
+    for line in order_lines:
+        variant = line.variant
+        if variant is None:
+            continue
+
+        channel_id = order.channel.id
```
Comment (fowczarek) [inline excise/utils.py:451]:
```suggestion
        channel_id = order.channel_id
```

File: excise/utils.py [hunk 8]
```diff
@@ -0,0 +1,554 @@
...
+        variant_channel_listing = line.variant.channel_listings.get(
+            channel_id=order.channel.id
```
Comment (fowczarek) [inline excise/utils.py:453]:
```suggestion
            channel_id=order.channel_id
```

**Comment by bufke** [comment]
Superseded by https://github.com/saleor/saleor-plugin-avatax-excise/pull/7

**Pull request closed without merging**